<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Смотрим-общую-информацию-и-чистим--данные" data-toc-modified-id="Смотрим-общую-информацию-и-чистим--данные-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Смотрим общую информацию и чистим  данные</a></span><ul class="toc-item"><li><span><a href="#Напишем-функцию-для-просмотра-общей-информации" data-toc-modified-id="Напишем-функцию-для-просмотра-общей-информации-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Напишем функцию для просмотра общей информации</a></span></li><li><span><a href="#Заменяем-пропущенные-значения" data-toc-modified-id="Заменяем-пропущенные-значения-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Заменяем пропущенные значения</a></span></li><li><span><a href="#Заменяем-отрицательные-значения" data-toc-modified-id="Заменяем-отрицательные-значения-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Заменяем отрицательные значения</a></span></li><li><span><a href="#Заменяем-аномальные-значения" data-toc-modified-id="Заменяем-аномальные-значения-3.1.4"><span class="toc-item-num">3.1.4&nbsp;&nbsp;</span>Заменяем аномальные значения</a></span></li><li><span><a href="#Заменяем-тип-данных" data-toc-modified-id="Заменяем-тип-данных-3.1.5"><span class="toc-item-num">3.1.5&nbsp;&nbsp;</span>Заменяем тип данных</a></span></li><li><span><a href="#Обрабатываем-дубликаты" data-toc-modified-id="Обрабатываем-дубликаты-3.1.6"><span class="toc-item-num">3.1.6&nbsp;&nbsp;</span>Обрабатываем дубликаты</a></span></li><li><span><a href="#Вызов-функции-df_info" data-toc-modified-id="Вызов-функции-df_info-3.1.7"><span class="toc-item-num">3.1.7&nbsp;&nbsp;</span>Вызов функции df_info</a></span></li></ul></li><li><span><a href="#Категоризуем-данные" data-toc-modified-id="Категоризуем-данные-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Категоризуем данные</a></span><ul class="toc-item"><li><span><a href="#Пишем-функцию-для-категоризации-клиентов-по-зарплатам" data-toc-modified-id="Пишем-функцию-для-категоризации-клиентов-по-зарплатам-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Пишем функцию для категоризации клиентов по зарплатам</a></span></li><li><span><a href="#Пишем-функцию-для-категоризации-целей-кредита" data-toc-modified-id="Пишем-функцию-для-категоризации-целей-кредита-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Пишем функцию для категоризации целей кредита</a></span></li></ul></li></ul></li><li><span><a href="#Исследовательский-анализ-данных" data-toc-modified-id="Исследовательский-анализ-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Исследовательский анализ данных</a></span><ul class="toc-item"><li><span><a href="#Узнаем-есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок" data-toc-modified-id="Узнаем-есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Узнаем есть ли зависимость между количеством детей и возвратом кредита в срок</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Узнаем-есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок" data-toc-modified-id="Узнаем-есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Узнаем есть ли зависимость между семейным положением и возвратом кредита в срок</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Узнаем-есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок" data-toc-modified-id="Узнаем-есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Узнаем есть ли зависимость между уровнем дохода и возвратом кредита в срок</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.3.1"><span class="toc-item-num">4.3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Узнаем-как-цели-на-кредит-влияют-на-возврат-в-срок" data-toc-modified-id="Узнаем-как-цели-на-кредит-влияют-на-возврат-в-срок-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Узнаем как цели на кредит влияют на возврат в срок</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.4.1"><span class="toc-item-num">4.4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод-по-исследованию" data-toc-modified-id="Общий-вывод-по-исследованию-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод по исследованию</a></span></li></ul></div>

# Описание проекта


Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

# Описание данных


- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

# Предобработка данных

*Импортируем необходимые библиотеки. Считаем данные из csv-файла в датафрейм и сохраним в переменную data*

In [1]:
import pandas as pd
from io import BytesIO
import requests

In [2]:
try:
    data = pd.read_csv('file:///Users/nastinpupochek/Downloads/data.csv')
except: 
    spreadsheet_id = '1qzoB2OFxsb-qiF1PUiIwmvzQ5WEj30jrGCKx84-PXe8'
    file_name = 'https://docs.google.com/spreadsheets/d/{}/export?format=csv'.format(spreadsheet_id)
    r = requests.get(file_name)
    data = pd.read_csv(BytesIO(r.content))

## Смотрим общую информацию и чистим  данные

###  Напишем функцию для просмотра общей информации

In [3]:
def df_info(df):
    display(df)
    print(f'Количество строк: {df.shape[0]}')
    print(f'\nКоличество пропущенных значений: \n{df.isna().sum()}')
    print(f'\nТипы данных в столбцах: \n{df.dtypes}')
    print(f'\nКоличество дубликатов: \n{df.duplicated().sum()}')
    return df_info
df_info(data)

children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

Количество строк: 21525

Количество пропущенных значений: 
children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Типы данных в столбцах: 
children              int64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

Количество дубликатов: 
54


<function __main__.df_info(df)>

### Заменяем пропущенные значения

*В двух столбцах есть пропущенные значения. days_employed и total_income. total_income — хранит данные о доходах, а days_employed — общий трудовой стаж в днях. На сумму дохода и сильнее всего влияет тип занятости, поэтому заполним пропуски в этом столбце  медианным значением по каждому типу из столбца income_type(тип занятости). days_employed так же заполним медианым значений по каждому типу из толбца income_type.*

*Заполняем total_income*

In [4]:
for med in data['income_type'].unique():
    data.loc[(data['income_type'] == med) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == med), 'total_income'].median()

*Заполняем days_employed*

In [5]:
for days in data['income_type'].unique():
    data.loc[(data['income_type'] == days) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == days), 'days_employed'].median()

### Заменяем отрицательные значения

*В столбце days_employed встречаются отрицательные значения, заменим эти значения на положительные с помощью метода abs.*

In [6]:
data['days_employed'].unique()

array([-8437.67302776, -4024.80375385, -5623.42261023, ...,
       -2113.3468877 , -3112.4817052 , -1984.50758853])

In [7]:
data.loc[data['days_employed'] < 0, 'days_employed'] = \
data.loc[data['days_employed'] < 0, 'days_employed'].abs()

In [8]:
data['days_employed'].unique()

array([8437.67302776, 4024.80375385, 5623.42261023, ..., 2113.3468877 ,
       3112.4817052 , 1984.50758853])

### Заменяем аномальные значения

*Проверим значения в столбце children. Удалим аномальные значения столбца, так как их малое количество*

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

In [10]:
data = data.loc[(data['children'] != -1) & (data['children'] != 20)]

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Заменяем тип данных

*Поменяем тип данных в столбце total_income на int*

In [12]:
data['total_income'] = data['total_income'].astype(int)

### Обрабатываем дубликаты

*Посмотрим на значения в столбце education*

In [14]:
data['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

In [15]:
data['education'] = data['education'].str.lower()

In [16]:
data['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

*У нас небольшое количество дубликатов, поэтому можем их спокойно удалить*

In [17]:
data = data.drop_duplicates()

### Вызов функции df_info

*Проверяем еще раз наш датасет, вызвав функцию df_info*

In [18]:
df_info(data)

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21520        224791             операции с жильем  
21521        155999          сделка с автомобилем  
21522         89672                  недвижимость  
21523        244093  на покупку своего автомобиля  
21524         82047         на покупку автомобиля  

[21331 rows x 12 columns]

Количество строк: 21331

Количество пропущенных значений: 
children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Типы данных в столбцах: 
children              int64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income          int64
purpose              object
dtype: object

Количество дубликатов: 
0


<function __main__.df_info(df)>

## Категоризуем данные

### Пишем функцию для категоризации клиентов по зарплатам

*Для удобства исследования, создадим категории заработка клиентов, где:*

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.*

In [19]:
def categorize_income(categor):
    if  0 <= categor <= 30000:
        return 'E'
    elif 30001 <= categor <= 50000:
        return 'D'
    elif 50001 <= categor <= 200000:
        return 'C'
    elif 200001 <= categor <= 1000000:
        return 'B'
    else:
        return 'A'

data['total_income_category'] = data['total_income'].apply(categorize_income)


### Пишем функцию для категоризации целей кредита

*На сновании данных из столбца purpose сформируем категории:*

- 'операции с автомобилем',
- 'операции с недвижимостью',
- 'проведение свадьбы',
- 'получение образования'.

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

In [21]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

# Исследовательский анализ данных

## Узнаем есть ли зависимость между количеством детей и возвратом кредита в срок

In [29]:
total = data['children'].value_counts()
child_count = data.pivot_table(index='children', 
                             columns='debt', 
                             values='purpose', 
                             aggfunc='count').reset_index()


In [33]:
child_count['debt_n'] = (child_count[1] / total * 100).round()

In [34]:
child_count['debt_y'] = (child_count[0] / total * 100).round()

In [35]:
child_count

debt  children        0       1  debt_y  debt_n
0            0  13028.0  1063.0     8.0    92.0
1            1   4364.0   444.0     9.0    91.0
2            2   1858.0   194.0     9.0    91.0
3            3    303.0    27.0     8.0    92.0
4            4     37.0     4.0    10.0    90.0
5            5      9.0     NaN     NaN   100.0

### Вывод

*Данные показывают, что процент возврата кредит в срок группы без детей, чуть выше. При этом при увеличенни количества детей в семье, невозврат кредита в срок на одном уровне. В семье с 5ью детьми стопроцентный вовзрат кредита.*

*Таким образом вне зависимости количества детей, кредит возращается в срок.*


## Узнаем есть ли зависимость между семейным положением и возвратом кредита в срок

*Узнаем процент невозврата кредита*

In [37]:
family_status = data.groupby('family_status')['debt'].sum() / \
      data.groupby('family_status')['debt'].count() * 100

In [39]:
family_status.round()

family_status
Не женат / не замужем    10.0
в разводе                 7.0
вдовец / вдова            7.0
гражданский брак          9.0
женат / замужем           8.0
Name: debt, dtype: float64

In [41]:
family_status_pivot = data.pivot_table(index=['family_status'], columns=['debt'], values='total_income', aggfunc='count')
family_status_pivot

debt                       0    1
family_status                    
Не женат / не замужем   2523  273
в разводе               1105   84
вдовец / вдова           888   63
гражданский брак        3749  385
женат / замужем        11334  927

### Вывод

*Разница в выплате кредитов среди групп с различным семейным положением низкая. Вдоцов/вдов с выплатой в срок на 1% больше женатых/замужем и на 3% больше, людей в гружданском браке. Люди в браке берут кредиты чаще, но при этом семейное положение на выплату кредита не влияет*

## Узнаем есть ли зависимость между уровнем дохода и возвратом кредита в срок

*Узнаем процент невозврата кредита*

In [43]:
monthly_income_set = data.groupby('total_income_category')['debt'].sum() / \
      data.groupby('total_income_category')['debt'].count() * 100

In [45]:
monthly_income_set.round()

total_income_category
A    8.0
B    7.0
C    8.0
D    6.0
E    9.0
Name: debt, dtype: float64

In [46]:
monthly_income_pivot = data.pivot_table(index=['total_income_category'], 
                                        columns=['debt'], 
                                        values='education', 
                                        aggfunc='count')


In [47]:
monthly_income_pivot

debt                       0     1
total_income_category             
A                         23     2
B                       4660   354
C                      14568  1353
D                        328    21
E                         20     2

### Вывод

*По полученным данным видно, что большинство кредитов берут люди из категории заработка В и С и у них высокие проценты по выплате в срок. Количество кредитов у категории D и Е намного ниже и при этом видно, что выплата кредита в срок низкая. При этом категория А скорее всего не берет кредиты сами, а категории Е отказывает банк.*

## Узнаем как цели на кредит влияют на возврат в срок

*Узнаем процент невозврата кредита*

In [51]:
purpose_income_set = data.groupby('purpose_category')['debt'].sum() / \
      data.groupby('purpose_category')['debt'].count() * 100

In [53]:
purpose_income_set.round()

purpose_category
операции с автомобилем      9.0
операции с недвижимостью    7.0
получение образования       9.0
проведение свадьбы          8.0
Name: debt, dtype: float64

In [49]:
credit_purpose_pivot = data.pivot_table(index=['purpose_category'], 
                                        columns=['debt'], 
                                        values='education', 
                                        aggfunc='count')

In [50]:
credit_purpose_pivot

debt                         0    1
purpose_category                   
операции с автомобилем    3879  400
операции с недвижимостью  9971  780
получение образования     3619  369
проведение свадьбы        2130  183

### Вывод

*По подсчитанным данным можно увидеть чаще всего кредиты берут на операции с недвижимостью, и возврат в срок равен 92%. На операции с автомобилем и получением образования выдалось примерно одинаковое количество кредитов, и выплаты в срок при этом тоже равны. Меньше всего бралось кредитов на проведение свадьбы, но выплата таких кредитов высока и равна 92%.*

# Общий вывод по исследованию

*Мы подсчитали 4 зависимости от различных групп к выплате кредита в срок по имеющиемся данным:*
* Выплата кредита в срок не зависит от количества детей.
* Зависимость выплаты кредита в срок от семейного положения низкая. Люди в браке чаще всего берут кредиты, но при это и выплата в срок высокая, как и в других группах
* Зависимость от уровня дохода и выплаты кредита в срок заметная. Люди с более низким доходом, категория D и E, выплачивают кредит в срок реже. Но при этом стоит учитывать, что чаще кредит выдавалася/брался в категориях дохода B и С. Выплаты в срок у этих категорий высокие. 
* Из целей кредита, больше всего бралось на операции с недвижимости, где и высокий возврат в срок. На операции с авто и получением образовнаия выдавалось практически равно количество кредитов, и выплаты в срок при этом тоже одинаковые. Меньше всего бралось кредитов на свадьбу, но при этом выплата кредита 92%. 



